gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java

```diff
@@ -139,13 +142,28 @@ public JsonWriter value(Number value) throws IOException {
     if (value == null) {
       return nullValue();
     }
-    double d = value.doubleValue();
     if (!isLenient()) {
```

Comment:
@bmazzarol sorry for being pedantic but what I meant is the following:

```java
   if (value instanceof Integer) {
      generator.writeNumber(value.intValue());
    } else if (value instanceof Short || value instanceof Byte) { // write byte as short
      generator.writeNumber(value.shortValue());
    } else if (value instanceof Long) {
      generator.writeNumber(value.longValue());
    } else if (value instanceof Float) {
      generator.writeNumber(value.floatValue());
   // } else if (value instanceof Double) {
   // remove instanceof Double. it is handled in else clause
   //  generator.writeNumber(value.doubleValue());
    } else if (value instanceof BigInteger) {
      generator.writeNumber((BigInteger) value);
    } else if (value instanceof BigDecimal) {
      generator.writeNumber((BigDecimal) value);
    } else {
       // by default, write as double
      // converting to double here (and only here)
     double d = value.doubleValue();
    if (!isLenient() && (Double.isNaN(d) || Double.isInfinite(d))) {
        throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value + " ( for " + value.getClass() + ")");	   
     }	
    generator.writeNumber(d);	    
    }
```

This way you convert to double only once.

Reply:
I know what you are asking for, but it is semantically different from the code you had before this pr, and is not correct. If I pass a Double value that is NaN or infinite and strict parsing is enabled no error will be raised.

Reply:
> If I pass a Double value that is NaN or infinite and strict parsing is enabled no error will be raised.

Would it not get into `else` statement and throw exception ? 

Or are you talking about the case when `BigDecimal.doubleValue()` returns `-Infinity` / `+Infinity` ?